manticore/wasm/structure.py

```diff
@@ -0,0 +1,1703 @@
+            FuncIdx
+        ] = None  #: https://www.w3.org/TR/wasm-core-1/#start-function%E2%91%A0
+        self.imports: typing.List[Import] = []
+        self.exports: typing.List[Export] = []
+
+    def __getstate__(self):
+        state = {
+            "types": self.types,
+            "funcs": self.funcs,
+            "tables": self.tables,
+            "mems": self.mems,
+            "globals": self.globals,
+            "elem": self.elem,
+            "data": self.data,
+            "start": self.start,
+            "imports": self.imports,
+            "exports": self.exports,
+            "_raw": self._raw,
+        }
+        return state
+
+    def __setstate__(self, state):
+        self.types = state["types"]
+        self.funcs = state["funcs"]
+        self.tables = state["tables"]
+        self.mems = state["mems"]
+        self.globals = state["globals"]
+        self.elem = state["elem"]
+        self.data = state["data"]
+        self.start = state["start"]
+        self.imports = state["imports"]
+        self.exports = state["exports"]
+        self._raw = state["_raw"]
+
+    @classmethod
+    def load(cls, filename: str):
+        """
+        Converts a WASM module in binary format into Python types that Manticore can understand
+
+        :param filename: name of the WASM module
+        :return: Module
+        """
+        type_map = {-16: types.FunctionType, -4: F64, -3: F32, -2: I64, -1: I32}
+
+        m: Module = cls()
+        with open(filename, "rb") as wasm_file:
+            m._raw = wasm_file.read()
+
+        module_iter = decode_module(m._raw)
+        _header = next(module_iter)
+        section: Section
+        # Parse the sections from the WASM module into internal types. For each section, the code usually resembles:
+        # `module.<something>.append(<InstanceOfSomething>(data, from, binary, module))`
+        for section, section_data in module_iter:
+            sec_id = section_data.id
+            if sec_id == SEC_TYPE:  # https://www.w3.org/TR/wasm-core-1/#type-section%E2%91%A0
+                for ft in section_data.payload.entries:
+                    m.types.append(
+                        FunctionType(
+                            [type_map[p_type] for p_type in ft.param_types],
+                            [type_map[ft.return_type] for _i in range(ft.return_count)],
+                        )
+                    )
+            elif sec_id == SEC_IMPORT:  # https://www.w3.org/TR/wasm-core-1/#import-section%E2%91%A0
+                for i in section_data.payload.entries:
+                    ty_map = i.get_decoder_meta()["types"]
+                    mod_name = strip_quotes(ty_map["module_str"].to_string(i.module_str))
+                    field_name = strip_quotes(ty_map["field_str"].to_string(i.field_str))
+                    if i.kind == 0:
+                        m.imports.append(Import(mod_name, field_name, TypeIdx(i.type.type)))
+
+                    elif i.kind == 1:
```

Comment:
Having a literal `1` here feels weird, but the specification does it too, so this is probably fine.